File: java/46_Permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(res, nums, 0);
        return res;
    }
    
    private void dfs(List<List<Integer>> res, int[] nums, int index) {
        if (index == nums.length) {
            List<Integer> temp = new ArrayList<>();
            for (int n : nums) {
                temp.add(n);
            }
            res.add(temp);
        }else {
            for (int i = index; i < nums.length; i++) {
                swap(nums, index, i);
                dfs(res, nums, index + 1);
                swap(nums, index, i);
            }
        }
    }
    
    private static void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: c++/46_Permutations.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> path;
    vector<bool> vis;
    
    vector<vector<int>> permute(vector<int>& nums) {
        path = vector<int>(nums.size());
        vis = vector<bool>(nums.size());
        dfs(nums, 0);
        return res;
    }
    
    void dfs(vector<int>& nums, int u) {
        if (u == nums.size()) {
            res.push_back(path);
        }
        for (int i = 0; i < nums.size(); i++) {
            if (!vis[i]) {
                path[u] = nums[i];
                vis[i] = true;
                dfs(nums, u+1);
                vis[i] = false;
            }
        }
    }
};

